sys/crypto/modes/ocb.c

```diff
@@ -0,0 +1,328 @@
+ * @{
+ *
+ * @file        ocb.c
+ * @brief       Offset Codebook (OCB3) AEAD mode as specified in RFC 7253
+ *
+ * @author      Mathias Tausig <mathias@tausig.at>
+ *
+ */
+
+#include "crypto/modes/ocb.h"
+#include <stdint.h>
+#include <string.h>
+
+#define OCB_MODE_ENCRYPT 1
+#define OCB_MODE_DECRYPT 2
+
+static void double_block(uint8_t source[16], uint8_t dest[16])
+{
+    uint8_t msb = source[0] >> 7;
+
+    for (uint8_t i = 0; i < 15; ++i) {
+        dest[i] = source[i] << 1 | source[i + 1] >> 7;
+    }
+    dest[15] = (source[15] << 1) ^ (0x87 * msb);
+}
+
+static size_t ntz(size_t n)
+{
+    /* ntz must only be run on positive values */
+    if (n == 0) {
+        return SIZE_MAX;
+    }
+
+    size_t ret = 0;
+
+    while (n % 2 == 0) {
+        ++ret;
+        n = n >> 1;
+    }
+    return ret;
+}
+
+static void calculate_l_i(uint8_t l_zero[16], size_t i, uint8_t output[16])
+{
+    memcpy(output, l_zero, 16);
+    while ((i--) > 0) {
+        double_block(output, output);
+    }
+}
+
+static void xor_block(uint8_t block1[16], uint8_t block2[16], uint8_t output[16])
+{
+    for (uint8_t i = 0; i < 16; ++i) {
+        output[i] = block1[i] ^ block2[i];
+    }
+}
+
+static void hash(cipher_t *cipher, uint8_t l_star[16], uint8_t l_zero[16],
+                 uint8_t *data, size_t data_len, uint8_t output[16])
+{
+
+    /* Calculate the number of full blocks in data */
+    size_t m = (data_len - (data_len % 16)) / 16;
+    size_t remaining_data_len = data_len - m * 16;
+
+
+    /* Sum_0 = zeros(128) */
+    memset(output, 0, 16);
+    /* Offset_0 = zeros(128) */
+    uint8_t offset[16];
+    memset(offset, 0, 16);
+    for (size_t i = 0; i < m; ++i) {
+        /* Offset_i = Offset_{i-1} xor L_{ntz(i)} */
+        uint8_t l_i[16];
+        calculate_l_i(l_zero, ntz(i + 1), l_i);
+        xor_block(offset, l_i, offset);
+        /* Sum_i = Sum_{i-1} xor ENCIPHER(K, A_i xor Offset_i) */
+        uint8_t enciphered_block[16], cipher_input[16];
+        xor_block(data, offset, cipher_input);
+        cipher->interface->encrypt(&(cipher->context), cipher_input, enciphered_block);
+        xor_block(output, enciphered_block, output);
+
+        data += 16;
+    }
+    if (remaining_data_len > 0) {
+        /* Offset_* = Offset_m xor L_* */
+        xor_block(offset, l_star, offset);
+        /* CipherInput = (A_* || 1 || zeros(127-bitlen(A_*))) xor Offset_* */
+        uint8_t cipher_input[16];
+        memset(cipher_input, 0, 16);
+        memcpy(cipher_input, data, remaining_data_len);
+        cipher_input[remaining_data_len] = 0x80;
+        xor_block(cipher_input, offset, cipher_input);
+        /* Sum = Sum_m xor ENCIPHER(K, CipherInput) */
+        uint8_t enciphered_block[16];
+        cipher->interface->encrypt(&(cipher->context), cipher_input, enciphered_block);
+        xor_block(output, enciphered_block, output);
+    }
+}
+
+static int32_t run_ocb(cipher_t *cipher, uint8_t *auth_data, uint32_t auth_data_len,
```

Comment:
This functions is too big. Please split it so it's easier to maintain (as stated in [Functions @ Coding Conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#functions))

Reply:
I optimized it with cd1c167, d120d55 and 09ae9de.
The function is still a bit over one screen page, but every shortening from this point would only be an artificial one (copying blocks into separate functions which are only called once).
I hope it is OK for you this way.

Reply:
yes, it's OK. Thanks!